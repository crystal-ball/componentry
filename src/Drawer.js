/* eslint-disable  react/prop-types */
import generateContainer from './toggleable/generate-container';
const Container = generateContainer('drawer');

/**
 * The `<Drawer>` component creates an expandable content drawer.
 *
 * @class Drawer
 * @constructor
 * @extends React.Component
 */
export default class Drawer extends Container {
  // Control arias used by subcomponents
  contentArias = { id: true }
  triggerArias = { controls: true, expanded: true }

  toggleActive = e => {
    let { active } = this.props;
    if (active === undefined) {
      active = this.state.active;
    }

    if (!active) {
      this.props.onActivate(this, e);
    } else {
      this.props.onDeactivate(this, e);
    }

    if (this.props.active === undefined) {
      // State is not controlled, update internal state
      this.setState({ active: !active });
    }

    if (!active) {
      this.props.onActivated(this, e);
    } else {
      this.props.onDeactivated(this, e);
    }
  }
}
