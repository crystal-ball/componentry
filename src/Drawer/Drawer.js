import activeContainer from '../factories/active-container-component'
import activeContent from '../factories/active-content-component'
import activeTrigger from '../factories/active-trigger-component'

/**
 * [Drawer component 📝](https://componentry.design/components/drawer)
 */
const Drawer = activeContainer('drawer')
export default Drawer

/**
 * [Drawer content component 📝](https://componentry.design/components/drawer)
 */
Drawer.Content = activeContent('drawer', { arias: { id: true, hidden: true } })

/**
 * [Drawer trigger component 📝](https://componentry.design/components/drawer)
 */
Drawer.Trigger = activeTrigger('drawer', { arias: { controls: true, expanded: true } })
