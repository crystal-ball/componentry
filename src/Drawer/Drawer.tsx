import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionComponent } from '../utils/active-action-component-builder'
import { activeContentComponent } from '../utils/active-content-component-builder'
import { BaseActiveContainerProps, BaseProps } from '../utils/base-types'

export interface DrawerProps
  extends BaseActiveContainerProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface DrawerActionProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DrawerContentProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Drawer {
  (props: DrawerProps): React.ReactElement
  /**
   * [Drawer action component üìù](https://componentry.design/components/drawer)
   */
  Action: React.FC<DrawerActionProps>
  /**
   * [Drawer content component üìù](https://componentry.design/components/drawer)
   */
  Content: React.FC<DrawerContentProps>
}

/**
 * [Drawer component üìù](https://componentry.design/components/drawer)
 */
export const Drawer = activeContainerBuilder<DrawerProps>('Drawer') as Drawer

Drawer.Action = activeActionComponent<DrawerActionProps>('DrawerAction', {
  aria: { controls: true, expanded: true },
})

Drawer.Content = activeContentComponent<DrawerContentProps>('DrawerContent', {
  aria: { id: true, hidden: true },
})
