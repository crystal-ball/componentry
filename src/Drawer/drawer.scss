/* componentry/src/Drawer/drawer */

.drawer-toggle,
.accordion-toggle {
  // Bs doesn't have real drawers, the 'Collapse' styles can be used to create
  // them, but b/c they don't include any structural styles they are not
  // convenient for creating standard drawers with minimal markup. Add display and
  // margin for standard drawers that could be easily extended/overriden
  display: block; // Prevent drawer content from starting at drawer button boundary
  margin-bottom: map-get($spacers, 2);
}
