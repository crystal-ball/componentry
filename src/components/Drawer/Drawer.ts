import { type ComponentPropsWithoutRef } from 'react'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  type ActiveActionBaseProps,
  type ActiveContainerBaseProps,
  type ActiveContentBaseProps,
} from '../../utils/base-types'
import { UtilityProps } from '../../utils/utility-classes'
import { Link } from '../Link/Link'

export interface DrawerProps
  extends ActiveContainerBaseProps,
    UtilityProps,
    ComponentPropsWithoutRef<'div'> {}

export interface DrawerActionProps
  extends ActiveActionBaseProps,
    UtilityProps,
    ComponentPropsWithoutRef<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DrawerContentProps
  extends ActiveContentBaseProps,
    UtilityProps,
    ComponentPropsWithoutRef<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Drawer {
  (props: DrawerProps): React.ReactElement
  /**
   * [Drawer action component üìù](https://componentry.design/components/drawer)
   */
  Action: React.FC<DrawerActionProps>
  /**
   * [Drawer content component üìù](https://componentry.design/components/drawer)
   */
  Content: React.FC<DrawerContentProps>
}

/**
 * [Drawer component üìù](https://componentry.design/components/drawer)
 * @experimental
 */
export const Drawer = activeContainerBuilder<DrawerProps>('Drawer') as Drawer

Drawer.Action = activeActionBuilder<DrawerActionProps>('DrawerAction', {
  aria: { controls: true, expanded: true },
  defaultAs: Link,
})

Drawer.Content = activeContentBuilder<DrawerContentProps>('DrawerContent', {
  aria: { id: true, hidden: true },
})
