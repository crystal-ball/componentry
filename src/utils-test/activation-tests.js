/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * This set of activation event tests should pass for every element with active
 * state. It's probably not really necessary to run these tests for each element,
 * but this provides a simple high level gut check on if functionality is working
 * as expected for each component.
 *
 * IMPORTANT NOTES
 * - `dt()` is a shorthand method for creating data-test strings. Using data-test
 *   strings lets us reuse shared tests across components
 * - Enzyme `find()` will find **EVERY** instance of a prop. Since we are passing
 *   data-test attributes down here, it will always find more than one. Use
 *   `.last()` to selecte final node.
 */

const findContent = wrapper => wrapper.find(dt('content')).last()
const findTrigger = wrapper => wrapper.find(dt('trigger')).last()

export default TestComponent => {
  /**
   * Test that by default Active.Trigger elements render as button nodes, and
   * when passed an href render anchor nodes.
   */
  test('should render button or anchor tags by default', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="btn" />
        <TestComponent.Trigger data-test="anchor" href="http://localhost:3000/">
          Anchor
        </TestComponent.Trigger>
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const btn = wrapper.find(dt('btn')).last()
    const anchor = wrapper.find(dt('anchor')).last()

    expect(anchor.type()).toEqual('a')
    expect(anchor.prop('type')).toEqual(undefined) // should not have type="button"
    expect(btn.type()).toEqual('button')
    expect(btn.prop('type')).toEqual('button') // should have type="button"
  })

  /**
   * Test that the default uncontrolled state scenario works. The active state should
   * be managed internally if not overriden with props and clicking the trigger should
   * show/hide content.
   */
  test('should update arias when trigger is activated', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    // Content
    const content = findContent(wrapper)
    const guid = content.prop('id')
    expect(content.length).toEqual(1)
    expect(content.prop('aria-hidden')).toEqual('true')

    // Trigger
    const trigger = findTrigger(wrapper)
    expect(trigger.length).toEqual(1)

    // TODO: need to figure out common aria attributes and test only those here
    // eg Popover and tooltip do not have aria-controls attrs
    if (trigger.prop('aria-controls')) expect(trigger.prop('aria-controls')).toEqual(guid)

    // Simulate a click, arias should update
    trigger.simulate('click')

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')

    // Simulate a click, arias should update
    trigger.simulate('click')

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
  })

  /**
   * Test that all of the user hooks are being called on activation/deactivation.
   * TODO: update to check that event hooks are called with correct args.
   */
  test('should call user event hooks', () => {
    const onActivate = jest.fn()
    const onActivated = jest.fn()
    const onDeactivate = jest.fn()
    const onDeactivated = jest.fn()
    const wrapper = mount(
      <TestComponent
        onActivate={onActivate}
        onActivated={onActivated}
        onDeactivate={onDeactivate}
        onDeactivated={onDeactivated}
      >
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const trigger = findTrigger(wrapper)

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1)
    expect(onActivated).toHaveBeenCalledTimes(1)
    expect(onDeactivate).toHaveBeenCalledTimes(0) // not yet!
    expect(onDeactivated).toHaveBeenCalledTimes(0) // not yet!

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1) // only once
    expect(onActivated).toHaveBeenCalledTimes(1) // only once
    expect(onDeactivate).toHaveBeenCalledTimes(1)
    expect(onDeactivated).toHaveBeenCalledTimes(1)
  })

  /**
   * Test setting initial active state using defaultActive prop
   */
  test('state should default to defaultActive when passed', () => {
    const wrapper = mount(
      <TestComponent defaultActive>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')
  })

  /**
   * State checks if active prop value is different from internal state and updates
   * if that's the case, but this should only happen when passed active is an actual
   * value and not undefined. Check this with an instance with no value, and a
   * forced rerender
   */
  test('should not use active prop when value is undefined', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
    wrapper.update()
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
  })

  /**
   * Test controlled component for methods `activate` and `deactivate`. Passing these
   * should override internal activate/deactivate meaning that clicking the trigger
   * for an element should only call those passed props.
   */
  test('should use passed activate and deactivate functions', () => {
    const activate = jest.fn()
    const deactivate = jest.fn()
    const wrapper = mount(
      <TestComponent activate={activate} deactivate={deactivate}>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const trigger = findTrigger(wrapper)

    trigger.simulate('click')

    // We have overridden the activate event so, dropdown should still be closed
    // and clicking trigger should call activate again

    expect(activate).toHaveBeenCalledTimes(1)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')

    // Mock controlled component being passed true, click should now ONLY call deactivate
    wrapper.setProps({ active: true })
    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(1)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(2)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')
  })

  /**
   * Test that any flavor of the `<Active />` component exposes internal state and
   * state change methods using the FaCC pattern
   */
  test('should expose internals using FaCC pattern', () => {
    const wrapper = mount(
      <TestComponent>
        {({ active, activate, deactivate }) => (
          <div>
            <span data-test="active">{String(active)}</span>
            <button type="button" onClick={activate} data-test="activate" />
            <button type="button" onClick={deactivate} data-test="deactivate" />
          </div>
        )}
      </TestComponent>,
    )

    expect(wrapper.find(dt('active')).text()).toEqual('false')
    wrapper.find(dt('activate')).simulate('click')
    expect(wrapper.find(dt('active')).text()).toEqual('true')
  })
}
