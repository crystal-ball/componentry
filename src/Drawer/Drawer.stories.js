/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'

import Drawer from './Drawer'

storiesOf('Drawer', module)
  .add('<Drawer />', () => (
    <Drawer>
      <Drawer.Trigger>Toggle drawer</Drawer.Trigger>
      <Drawer.Content>Drawer content</Drawer.Content>
    </Drawer>
  ))
  .add('Collection', () => (
    <div>
      <h2>Drawer</h2>
      <div>
        <Drawer>
          <Drawer.Trigger>Toggle drawer</Drawer.Trigger>
          <Drawer.Content>Drawer content</Drawer.Content>
        </Drawer>
      </div>

      <h2>Multiple drawers</h2>
      <div>
        <Drawer>
          <Drawer.Trigger activeId='one'>Toggle drawer</Drawer.Trigger>
          <Drawer.Content activeId='one'>Drawer content</Drawer.Content>

          <Drawer.Trigger activeId='two'>Toggle drawer</Drawer.Trigger>
          <Drawer.Content activeId='two'>Drawer content</Drawer.Content>

          <Drawer.Trigger activeId='three'>Toggle drawer</Drawer.Trigger>
          <Drawer.Content activeId='three'>Drawer content</Drawer.Content>
        </Drawer>
      </div>
    </div>
  ))
