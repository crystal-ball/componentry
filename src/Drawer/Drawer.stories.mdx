import { Meta } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Drawer } from './Drawer'

<Meta
  title='Components/Drawer'
  component={Drawer}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Drawer

The Drawer component allows showing and hiding content

<Drawer>
  <Drawer.Action>Toggle drawer</Drawer.Action>
  <Drawer.Content>Drawer content</Drawer.Content>
</Drawer>

## Drawer

<Drawer>
  <Drawer.Action>Toggle drawer</Drawer.Action>
  <Drawer.Content>Drawer content</Drawer.Content>
</Drawer>

## Multiple drawers

<Drawer>
  <Drawer.Action activeId='one'>Toggle drawer</Drawer.Action>
  <Drawer.Content activeId='one'>Drawer content</Drawer.Content>
  <Drawer.Action activeId='two'>Toggle drawer</Drawer.Action>
  <Drawer.Content activeId='two'>Drawer content</Drawer.Content>
  <Drawer.Action activeId='three'>Toggle drawer</Drawer.Action>
  <Drawer.Content activeId='three'>Drawer content</Drawer.Content>
</Drawer>
