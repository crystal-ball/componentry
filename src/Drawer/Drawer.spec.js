import React from 'react'
import { render } from '@testing-library/react'

import { activationTests } from '../../test/activation-tests'
import { elementTests } from '../../test/element-tests'
import { Drawer } from './Drawer'

describe('<Drawer />', () => {
  activationTests(Drawer, { name: 'drawer', testArias: ['controls', 'expanded'] })

  elementTests(Drawer)
  elementTests(Drawer.Trigger)
  elementTests(Drawer.Content)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Drawer /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Drawer>
        <Drawer.Trigger>Trigger</Drawer.Trigger>
        <Drawer.Content>Content</Drawer.Content>
      </Drawer>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
