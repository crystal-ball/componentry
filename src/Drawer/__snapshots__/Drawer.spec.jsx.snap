// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> snapshots renders correctly 1`] = `
<div
  class="drawer"
  data-id="guid"
  data-test="drawer-container"
>
  <button
    aria-controls="guid"
    aria-expanded="false"
    class="btn drawer-toggle"
    data-test="drawer-trigger"
    type="button"
  >
    Trigger
  </button>
  <div
    aria-hidden="true"
    class="drawer-content"
    data-test="drawer-content"
    id="guid"
  >
    Content
  </div>
</div>
`;
